<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CiteFuse • BiocAsia2020CiteFuse</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CiteFuse">
<meta property="og:description" content="BiocAsia2020CiteFuse">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocAsia2020CiteFuse</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CiteFuse_BioCAsia_workshop.html">CiteFuse</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SydneyBioX/BiocAsia2020CiteFuse">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="CiteFuse_BioCAsia_workshop_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="CiteFuse_BioCAsia_workshop_files/jquery-1.12.4/jquery.min.js"></script><link href="CiteFuse_BioCAsia_workshop_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="CiteFuse_BioCAsia_workshop_files/datatables-binding-0.13/datatables.js"></script><link href="CiteFuse_BioCAsia_workshop_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="CiteFuse_BioCAsia_workshop_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="CiteFuse_BioCAsia_workshop_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="CiteFuse_BioCAsia_workshop_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="CiteFuse_BioCAsia_workshop_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CiteFuse</h1>
                        <h4 class="author">Yingxin Lin<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> and Hani Jieun Kim<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/BiocAsia2020CiteFuse/blob/master/vignettes/CiteFuse_BioCAsia_workshop.Rmd"><code>vignettes/CiteFuse_BioCAsia_workshop.Rmd</code></a></small>
      <div class="hidden name"><code>CiteFuse_BioCAsia_workshop.Rmd</code></div>

    </div>

    
    
<div id="workshop-title" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-title" class="anchor"></a>Workshop Title</h1>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<div id="instructors-and-contact-information" class="section level3">
<h3 class="hasAnchor">
<a href="#instructors-and-contact-information" class="anchor"></a>Instructors and contact information</h3>
<ul>
<li>Yingxin Lin (<a href="mailto:yingxin.lin@sydney.edu.au" class="email">yingxin.lin@sydney.edu.au</a>)</li>
<li>Hani Jieun Kim (<a href="mailto:hani.kim@sydney.edu.au" class="email">hani.kim@sydney.edu.au</a>)</li>
</ul>
</div>
<div id="description" class="section level3">
<h3 class="hasAnchor">
<a href="#description" class="anchor"></a>Description</h3>
<p>The latest breakthrough in single-cell omics is on multi-modal profiling of different biomolecule species in individual cells. Among the fast evolving biotechnologies developed for multi-modal profiling, cellular indexing of transcriptomes and epitopes by sequencing (CITE-seq) is attracting attention, especially in the field of immunology, given its ability to simultaneously quantify global gene expression and cellular proteins using RNA-sequencing and antibody-derived tags (ADTs) on single cells. While the additional protein marker information brought about by ADTs is extremely valuable, new biological insights can only be gained by developing analytic methods that fully take advantage of the complementarity between mRNA and ADT expression measured in CITE-seq.</p>
<p>To address this, we developed a streamlined pipeline–CiteFuse–that consists of a suite of tools for the integration and the downstream analysis of CITE-seq data. In this workshop, we will provide a hands-on experience to the CiteFuse package and cover all the steps including doublet detection, modality integration, cell type clustering, differential RNA and ADT expression analysis, ADT evaluation, and ligand–receptor interaction analysis on a publicly available CITE-seq dataset. We also demonstrate the applicability of CiteFuse package on other multi-modal data types by applying our pipeline on the recently developed ASAP-seq data.</p>
<p><a href="https://sydneybiox.github.io/CiteFuse/articles/CiteFuse.html">This vignette</a> provides a more complete description of the various tools in CiteFuse and will serve as the basis of our workshop.</p>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<p>Software:</p>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with single-cell RNA-sequencing</li>
<li>Familiarity with the <code>SingleCellExperiment</code> class</li>
</ul>
<p>Background reading:</p>
<ul>
<li>The textbook “Orchestrating Single-Cell Analysis with Bioconductor” is a great reference for single-cell analysis using Bioconductor packages.</li>
<li><a href="https://academic.oup.com/bioinformatics/article-abstract/36/14/4137/5827474?redirectedFrom=fulltext">CiteFuse enables multi-modal analysis of CITE-seq data</a></li>
<li><a href="https://www.nature.com/articles/nmeth.4380">Simultaneous epitope and transcriptome measurement in single cells</a></li>
</ul>
</div>
<div id="participation" class="section level3">
<h3 class="hasAnchor">
<a href="#participation" class="anchor"></a>Participation</h3>
<p>Participants will learn how to…</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li>This workshop will focus on Bioconductor packages [SingleCellExperiment] (<a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="uri">https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html</a>) and <a href="https://academic.oup.com/bioinformatics/article-abstract/36/14/4137/5827474?redirectedFrom=fulltext">CiteFuse</a>.</li>
</ul>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<p>An example for a 55-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Data processing and integration</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Downstream analysis</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Application of CiteFuse on ASAP-seq</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Wrap-up and Conclusions</td>
<td>5m</td>
</tr>
</tbody>
</table>
</div>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h3>
<p>Some examples:</p>
<ul>
<li>describe how to…</li>
<li>identify methods for…</li>
<li>understand the difference between…</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h3>
<ul>
<li>analyze xyz data to produce…</li>
<li>create xyz plots</li>
<li>evaluate xyz data for artifacts</li>
</ul>
</div>
</div>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<div id="the-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#the-dataset" class="anchor"></a>The dataset</h2>
<p>The dataset we will be working with concerns a single-cell RNA-sequencing dataset consisting of two different experiments, which correspond to two treatments. <a href="https://www.nature.com/articles/s41588-019-0489-5">McFaline-Figueroa et al.</a> studied the epithelial-to-mesenchymal transition (EMT), where cells spatially migrate from the epithelium to the mesenchyme during development. This process will be described by a trajectory, reflecting the gene expression changes occurring during this migration. The authors furthermore studied both a control (<code>Mock</code>) condition, and a condition in the absence of transforming growth factor <span class="math inline">\(\beta\)</span> (TGFB).</p>
<p>In summary, we will be predominantly investigating a CITE-seq dataset consisting of PBMCs. For simplicity, this lineage is studied in two different conditions; a control condition and a TGFB-absent condition.</p>
<p>First, install <code>CiteFuse</code> using <code>BiocManager</code>.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span> <span class="kw">=</span> <span class="fl">TRUE</span>)) {
 <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
}
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"CiteFuse"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">CiteFuse</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DT</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">igraph</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">mclust</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">gridExtra</span>)</pre></body></html></div>
<p>We have made this public dataset available as a <code>SingleCellExperiment</code> object in the workshop package, which we import below.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"CITEseq_example"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"CiteFuse"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">CITEseq_example</span>)
<span class="co">#&gt; [1] "RNA" "ADT" "HTO"</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">CITEseq_example</span>, <span class="no">dim</span>)
<span class="co">#&gt; $RNA</span>
<span class="co">#&gt; [1] 19521   500</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ADT</span>
<span class="co">#&gt; [1]  49 500</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $HTO</span>
<span class="co">#&gt; [1]   4 500</span></pre></body></html></div>
<p>Here, we start from a list of three matrices of unique molecular identifier (UMI), antibody derived tags (ADT) and hashtag oligonucleotide (HTO) count, which have common cell names. There are 500 cells in our subsetted dataset. And characteristically of CITE-seq data, the matrices are matched, meaning that for any given cell we know the expression level of their RNA transcripts (genome-wide) and its corresponding cell surface protein expression. The <code>preprocessing</code> function will utilise the three matrices and its common cell names to create a <code>SingleCellExperiment</code> object, which stores RNA data in an <code>assay</code> and <code>ADT</code> and <code>HTO</code> data within in the <code>altExp</code> slot.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/preprocessing.html">preprocessing</a></span>(<span class="no">CITEseq_example</span>)
<span class="no">sce_citeseq</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 19521 500 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(19521): hg19_AL627309.1 hg19_AL669831.5 ... hg19_MT-ND6</span>
<span class="co">#&gt;   hg19_MT-CYB</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames(500): AAGCCGCGTTGTCTTT GATCGCGGTTATCGGT ... TTGGCAACACTAGTAC</span>
<span class="co">#&gt;   GCTGCGAGTTGTGGCC</span>
<span class="co">#&gt; colData names(0):</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(2): ADT HTO</span></pre></body></html></div>
</div>
<div id="doublet-detection-using-citefuse" class="section level2">
<h2 class="hasAnchor">
<a href="#doublet-detection-using-citefuse" class="anchor"></a>Doublet detection using <code>CiteFuse</code>
</h2>
<div id="hto-normalisation-and-visualisation" class="section level3">
<h3 class="hasAnchor">
<a href="#hto-normalisation-and-visualisation" class="anchor"></a>HTO Normalisation and Visualisation</h3>
<p>The function <code>normaliseExprs</code> is used to scale the alternative expression. Here, we used it to perform log-transformation of the <code>HTO</code> count, by setting <code>transform = "log"</code>.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html">normaliseExprs</a></span>(<span class="kw">sce</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>,
                              <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"HTO"</span>,
                              <span class="kw">transform</span> <span class="kw">=</span> <span class="st">"log"</span>)</pre></body></html></div>
<p>Then we can perform dimension reduction on the <code>HTO</code> count by using <code>runTSNE</code> or <code>runUMAP</code>, then use <code>visualiseDim</code> function to visualise the reduced dimension plot. Our CITE-seq dataset contain data from four samples that were pooled before sequencing. The samples were multiplexed through cell hashing (Stoekius et al., 2018). The four clusters observed on reduced dimension plots equate to the four different samples.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="kw pkg">scater</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE</a></span>(<span class="no">sce_citeseq</span>,
                               <span class="kw">altexp</span> <span class="kw">=</span> <span class="st">"HTO"</span>,
                               <span class="kw">name</span> <span class="kw">=</span> <span class="st">"TSNE_HTO"</span>,
                               <span class="kw">pca</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>,
             <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"TSNE_HTO"</span>) + <span class="fu">labs</span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"tSNE (HTO)"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-7-1.png" width="576"></p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="kw pkg">scater</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span>(<span class="no">sce_citeseq</span>,
                               <span class="kw">altexp</span> <span class="kw">=</span> <span class="st">"HTO"</span>,
                               <span class="kw">name</span> <span class="kw">=</span> <span class="st">"UMAP_HTO"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>,
             <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"UMAP_HTO"</span>) + <span class="fu">labs</span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"UMAP (HTO)"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-7-2.png" width="576"></p>
</div>
<div id="doublet-identification-step-1-cross-sample-doublet-detection" class="section level3">
<h3 class="hasAnchor">
<a href="#doublet-identification-step-1-cross-sample-doublet-detection" class="anchor"></a>Doublet identification step 1: cross-sample doublet detection</h3>
<p>An important step in single cell data analysis is the removal of doublets. Doublets form as a result of co-encapsulation of cells within a droplet, leading to a hybrid transcriptome from two or more cells. In CiteFuse, we implement a step-wise doublet detection approach to remove doublets. We first identify the cross-sample doublets via the <code>crossSampleDoublets</code> function.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/crossSampleDoublets.html">crossSampleDoublets</a></span>(<span class="no">sce_citeseq</span>)
<span class="co">#&gt; number of iterations= 20 </span>
<span class="co">#&gt; number of iterations= 24 </span>
<span class="co">#&gt; number of iterations= 46 </span>
<span class="co">#&gt; number of iterations= 50</span></pre></body></html></div>
<p>The results of the cross sample doublets are then saved in <code>colData</code> as <code>doubletClassify_between_label</code> and <code>doubletClassify_between_class</code>.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">sce_citeseq</span>$<span class="no">doubletClassify_between_label</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;                 1                 2                 3                 4 </span>
<span class="co">#&gt;               115               121                92               129 </span>
<span class="co">#&gt; doublet/multiplet </span>
<span class="co">#&gt;                43</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">sce_citeseq</span>$<span class="no">doubletClassify_between_class</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; doublet/multiplet           Singlet </span>
<span class="co">#&gt;                43               457</span></pre></body></html></div>
<p>We can then highlight the cross-sample doublets in our tSNE plot of HTO count.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>,
             <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"TSNE_HTO"</span>,
             <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"doubletClassify_between_label"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-10-1.png" width="576"></p>
<p>Furthermore, <code>plotHTO</code> function allows us to plot the pairwise scatter HTO count. Any cells that show co-expression of orthologocal HTOs (red) are considered as doublets.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/plotHTO.html">plotHTO</a></span>(<span class="no">sce_citeseq</span>, <span class="fl">1</span>:<span class="fl">4</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-11-1.png" width="576"></p>
</div>
<div id="doublet-identification-step-1-within-sample-doublet-detection" class="section level3">
<h3 class="hasAnchor">
<a href="#doublet-identification-step-1-within-sample-doublet-detection" class="anchor"></a>Doublet identification step 1: within-sample doublet detection</h3>
<p>We then identify the within-sample doublets via the <code>withinSampleDoublets</code> function.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/withinSampleDoublets.html">withinSampleDoublets</a></span>(<span class="no">sce_citeseq</span>,
                                    <span class="kw">minPts</span> <span class="kw">=</span> <span class="fl">10</span>)</pre></body></html></div>
<p>The results of the cross sample doublets are then saved in the <code>colData</code> as <code>doubletClassify_within_label</code> and <code>doubletClassify_within_class</code>.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">sce_citeseq</span>$<span class="no">doubletClassify_within_label</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;  Doublets(Within)_1  Doublets(Within)_2  Doublets(Within)_3  Doublets(Within)_4 </span>
<span class="co">#&gt;                   3                   7                   4                   6 </span>
<span class="co">#&gt; NotDoublets(Within) </span>
<span class="co">#&gt;                 480</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">sce_citeseq</span>$<span class="no">doubletClassify_within_class</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; Doublet Singlet </span>
<span class="co">#&gt;      20     480</span></pre></body></html></div>
<p>Again, we can visualise the within-sample doublets in our tSNE plot.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>,
             <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"TSNE_HTO"</span>,
             <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"doubletClassify_within_label"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-14-1.png" width="576"></p>
<p>Finally, we can filter out the doublet cells (both within and between batches) for the downstream analysis.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="no">sce_citeseq</span>[, <span class="no">sce_citeseq</span>$<span class="no">doubletClassify_within_class</span> <span class="kw">==</span> <span class="st">"Singlet"</span> <span class="kw">&amp;</span> <span class="no">sce_citeseq</span>$<span class="no">doubletClassify_between_class</span> <span class="kw">==</span> <span class="st">"Singlet"</span>]
<span class="no">sce_citeseq</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 19521 437 </span>
<span class="co">#&gt; metadata(3): doubletClassify_between_threshold</span>
<span class="co">#&gt;   doubletClassify_between_resultsMat doubletClassify_within_resultsMat</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(19521): hg19_AL627309.1 hg19_AL669831.5 ... hg19_MT-ND6</span>
<span class="co">#&gt;   hg19_MT-CYB</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames(437): GATCGCGGTTATCGGT GGCTGGTAGAGGTTAT ... TTGGCAACACTAGTAC</span>
<span class="co">#&gt;   GCTGCGAGTTGTGGCC</span>
<span class="co">#&gt; colData names(5): doubletClassify_between_label</span>
<span class="co">#&gt;   doubletClassify_between_class nUMI doubletClassify_within_label</span>
<span class="co">#&gt;   doubletClassify_within_class</span>
<span class="co">#&gt; reducedDimNames(2): TSNE_HTO UMAP_HTO</span>
<span class="co">#&gt; altExpNames(2): ADT HTO</span></pre></body></html></div>
</div>
</div>
<div id="clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering</h2>
<div id="performing-snf" class="section level3">
<h3 class="hasAnchor">
<a href="#performing-snf" class="anchor"></a>Performing SNF</h3>
<p>The first step of analysis is to integrate the RNA and ADT matrix. We use a popular integration algorithm called similarity network fusion (SNF) to integrate the multiomic data.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="kw pkg">scater</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/logNormCounts.html">logNormCounts</a></span>(<span class="no">sce_citeseq</span>)
<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html">normaliseExprs</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>, <span class="kw">transform</span> <span class="kw">=</span> <span class="st">"log"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/CiteFuse.html">CiteFuse</a></span>(<span class="no">sce_citeseq</span>))
<span class="co">#&gt; Calculating affinity matrix </span>
<span class="co">#&gt; Performing SNF</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   5.320   0.144   5.471</span></pre></body></html></div>
<p>We now proceed with the fused matrix, which is stored as <code>SNF_W</code> in our <code>sce_citeseq</code> object.</p>
</div>
<div id="performing-spectral-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#performing-spectral-clustering" class="anchor"></a>Performing spectral clustering</h3>
<p>CiteFuse implements two different clustering algorithms on the fused matrix, spectral clustering and Louvain clustering. First, we perform spectral clustering with sufficient numbers of <code>K</code> and use the eigen values to determine the optimal number of clusters.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">SNF_W_clust</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html">spectralClustering</a></span>(<span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"SNF_W"</span>]], <span class="kw">K</span> <span class="kw">=</span> <span class="fl">20</span>)
<span class="co">#&gt; Computing Spectral Clustering</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">SNF_W_clust</span>$<span class="no">eigen_values</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/diff.html">diff</a></span>(<span class="no">SNF_W_clust</span>$<span class="no">eigen_values</span>)))
<span class="co">#&gt; [1] 5</span></pre></body></html></div>
<p>Using the optimal cluster number defined from the previous step, we can now use the <code>spectralClutering</code> function to cluster the single cells by specifying the number of clusters in <code>K</code>. The function takes a cell-to-cell similarity matrix as an input. We have already created the fused similarity matrix from <code>CiteFuse</code>. Since the <code>CiteFuse</code> function creates and stores the similarity matries from ADT and RNA expression, as well the fused matrix, we can use these two to compare the clustering outcomes by data modality.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">SNF_W_clust</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html">spectralClustering</a></span>(<span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"SNF_W"</span>]], <span class="kw">K</span> <span class="kw">=</span> <span class="fl">5</span>)
<span class="co">#&gt; Computing Spectral Clustering</span>
<span class="no">sce_citeseq</span>$<span class="no">SNF_W_clust</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">SNF_W_clust</span>$<span class="no">labels</span>)
<span class="no">SNF_W1_clust</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html">spectralClustering</a></span>(<span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"ADT_W"</span>]], <span class="kw">K</span> <span class="kw">=</span> <span class="fl">5</span>)
<span class="co">#&gt; Computing Spectral Clustering</span>
<span class="no">sce_citeseq</span>$<span class="no">ADT_clust</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">SNF_W1_clust</span>$<span class="no">labels</span>)
<span class="no">SNF_W2_clust</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html">spectralClustering</a></span>(<span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"RNA_W"</span>]], <span class="kw">K</span> <span class="kw">=</span> <span class="fl">5</span>)
<span class="co">#&gt; Computing Spectral Clustering</span>
<span class="no">sce_citeseq</span>$<span class="no">RNA_clust</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">SNF_W2_clust</span>$<span class="no">labels</span>)</pre></body></html></div>
</div>
<div id="visualisation" class="section level3">
<h3 class="hasAnchor">
<a href="#visualisation" class="anchor"></a>Visualisation</h3>
<p>The outcome of the clustering can be easily visualised on a reduced dimensions plot by highlighting the points by cluster label.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/reducedDimSNF.html">reducedDimSNF</a></span>(<span class="no">sce_citeseq</span>,
                             <span class="kw">method</span> <span class="kw">=</span> <span class="st">"tSNE"</span>,
                             <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"tSNE_joint"</span>)
<span class="no">g1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"tSNE_joint"</span>, <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"SNF_W_clust"</span>) +
  <span class="fu">labs</span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"tSNE (SNF clustering)"</span>)
<span class="no">g2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"tSNE_joint"</span>,  <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"ADT_clust"</span>) +
  <span class="fu">labs</span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"tSNE (ADT clustering)"</span>)
<span class="no">g3</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"tSNE_joint"</span>,  <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"RNA_clust"</span>) +
  <span class="fu">labs</span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"tSNE (RNA clustering)"</span>)

<span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html">grid.arrange</a></span>(<span class="no">g3</span>, <span class="no">g2</span>, <span class="no">g1</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-19-1.png" width="768"></p>
<p>The expression of genes and proteins can be visualised by changing the <code>colour_by</code> parameter to assess the clusters. As an example, we highlight the plot by the RNA and ADT expression level of CD8.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">g1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"tSNE_joint"</span>,
                   <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"hg19_CD8A"</span>,
                   <span class="kw">data_from</span> <span class="kw">=</span> <span class="st">"assay"</span>,
                   <span class="kw">assay_name</span> <span class="kw">=</span> <span class="st">"logcounts"</span>) +
  <span class="fu">labs</span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"tSNE: hg19_CD8A (RNA expression)"</span>)
<span class="no">g2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>,<span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"tSNE_joint"</span>,
                   <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"CD8"</span>,
                   <span class="kw">data_from</span> <span class="kw">=</span> <span class="st">"altExp"</span>,
                   <span class="kw">altExp_assay_name</span> <span class="kw">=</span> <span class="st">"logcounts"</span>) +
  <span class="fu">labs</span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"tSNE: CD8 (ADT expression)"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html">grid.arrange</a></span>(<span class="no">g1</span>, <span class="no">g2</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
</div>
<div id="louvain-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#louvain-clustering" class="anchor"></a>Louvain clustering</h3>
<p>As well as spectral clustering, CiteFuse can implement Louvain clustering if users wish to use another clustering method. We use the <code>igraph</code> package, and any community detection algorithms available in their package can be selected by changing the <code>method</code> parameter.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">SNF_W_louvain</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/igraphClustering.html">igraphClustering</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"louvain"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">SNF_W_louvain</span>)
<span class="co">#&gt; SNF_W_louvain</span>
<span class="co">#&gt;  1  2  3  4  5  6  7  8  9 </span>
<span class="co">#&gt; 30 52 42 31 72 56 37 88 29</span>
<span class="no">sce_citeseq</span>$<span class="no">SNF_W_louvain</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">SNF_W_louvain</span>)
<span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html">visualiseDim</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">dimNames</span> <span class="kw">=</span> <span class="st">"tSNE_joint"</span>, <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>) +
  <span class="fu">labs</span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"tSNE (SNF louvain clustering)"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-21-1.png" width="768"></p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseKNN.html">visualiseKNN</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">colour_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-22-1.png" width="576"></p>
</div>
</div>
<div id="differential-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-analysis" class="anchor"></a>Differential Expression Analysis</h2>
<div id="exploration-of-feature-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#exploration-of-feature-expression" class="anchor"></a>Exploration of feature expression</h3>
<p>CiteFuse has a wide range of visualisation tools to facilitate exploratory analysis of CITE-seq data. The <code>visualiseExprs</code> function is an easy-to-use function to generate boxplots, violinplots, jitter plots, density plots, and pairwise scatter/density plots of genes and proteins expressed in the data. The plots can be grouped by using the cluster labels stored in the <code>sce_citeseq</code> object.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"boxplot"</span>,
               <span class="kw">group_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>,
               <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-23-1.png" width="768"></p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"violin"</span>,
               <span class="kw">group_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>,
               <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-23-2.png" width="768"></p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"jitter"</span>,
               <span class="kw">group_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>,
               <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-23-3.png" width="768"></p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"density"</span>,
               <span class="kw">group_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>,
               <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-23-4.png" width="768"></p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>,
               <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
               <span class="kw">group_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"violin"</span>, <span class="kw">n</span> <span class="kw">=</span> <span class="fl">5</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-24-1.png" width="576"></p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"jitter"</span>,
               <span class="kw">group_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>,
               <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CD2"</span>, <span class="st">"CD8"</span>, <span class="st">"CD4"</span>, <span class="st">"CD19"</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-24-2.png" width="576"></p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"density"</span>,
               <span class="kw">group_by</span> <span class="kw">=</span> <span class="st">"SNF_W_louvain"</span>,
               <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CD2"</span>, <span class="st">"CD8"</span>, <span class="st">"CD4"</span>, <span class="st">"CD19"</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-24-3.png" width="576"></p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"pairwise"</span>,
               <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CD4"</span>, <span class="st">"CD8"</span>))
<span class="co">#&gt; number of iterations= 12 </span>
<span class="co">#&gt; number of iterations= 18</span></pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-25-1.png" width="768"></p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html">visualiseExprs</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
               <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"pairwise"</span>,
               <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CD45RA"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span>),
               <span class="kw">threshold</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">4</span>, <span class="fl">3</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-25-2.png" width="768"></p>
</div>
<div id="perform-de-analysis-with-wilcoxon-rank-sum-test" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-de-analysis-with-wilcoxon-rank-sum-test" class="anchor"></a>Perform DE Analysis with Wilcoxon Rank Sum test</h3>
<p>CiteFuse also calculates differentially expressed (DE) genes through the <code>DEgenes</code> function. The cluster grouping to use must be specified in the <code>group</code> parameter. If <code>altExp_name</code> is not specified, RNA expression will be used as the default expression matrix.</p>
<p>Results form the DE analysis is stored in <code>sce_citeseq</code> as <code>DE_res_RNA_filter</code> and <code>DE_res_ADT_filter</code> for RNA and ADT expression, respectively.</p>
</div>
<div id="for-rna-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#for-rna-expression" class="anchor"></a>For RNA expression</h3>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="co"># DE will be performed for RNA if altExp_name = "none" </span>
<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEgenes.html">DEgenes</a></span>(<span class="no">sce_citeseq</span>,
                       <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"none"</span>,
                       <span class="kw">group</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>$<span class="no">SNF_W_louvain</span>,
                       <span class="kw">return_all</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                       <span class="kw">exprs_pct</span> <span class="kw">=</span> <span class="fl">0.5</span>)
<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/selectDEgenes.html">selectDEgenes</a></span>(<span class="no">sce_citeseq</span>,
                             <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"none"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/format.html">format</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="no">rbind</span>, <span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"DE_res_RNA_filter"</span>]]),
                 <span class="kw">digits</span> <span class="kw">=</span> <span class="fl">2</span>))</pre></body></html></div>
<div id="htmlwidget-8718cccf694393e48273" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-8718cccf694393e48273">{"x":{"filter":"none","data":[["1.hg19_GZMK","1.hg19_KLRB1","1.hg19_DUSP2","1.hg19_IL7R","1.hg19_NCR3","1.hg19_CEBPD","1.hg19_KLRG1","1.hg19_CD69","1.hg19_LINC01871","1.hg19_IL32","2.hg19_SH2D1B","2.hg19_KLRF1","2.hg19_CLIC3","2.hg19_FCER1G","2.hg19_MYOM2","2.hg19_TRDC","2.hg19_SPON2","2.hg19_TYROBP","2.hg19_IL2RB","2.hg19_KLRB1","3.hg19_CD8B","3.hg19_CCL5","3.hg19_CD8A","3.hg19_CMC1","3.hg19_KLRG1","3.hg19_DUSP2","3.hg19_FYN","3.hg19_GZMM","3.hg19_SRP14","3.hg19_HCST","4.hg19_IL7R","4.hg19_S100A4","4.hg19_S100A10","4.hg19_IL32","4.hg19_FAM107B","4.hg19_EML4","4.hg19_CD6","4.hg19_CD2","4.hg19_RORA","4.hg19_ITGB1","5.hg19_TRBV20-1","5.hg19_ZNF683","5.hg19_FGFBP2","5.hg19_CD8A","5.hg19_GZMH","5.hg19_CD8B","5.hg19_CYTOR","5.hg19_CST7","5.hg19_KLRD1","5.hg19_CCL5","6.hg19_GNLY","6.hg19_CCL5","6.hg19_CTSW","6.hg19_GZMB","6.hg19_GZMH","6.hg19_IFITM2","6.hg19_NKG7","6.hg19_PRSS23","6.hg19_CD8A","6.hg19_HOPX","7.hg19_FCN1","7.hg19_LYZ","7.hg19_SPI1","7.hg19_CYBB","7.hg19_HLA-DRA","7.hg19_S100A9","7.hg19_CD68","7.hg19_S100A8","7.hg19_CSTA","7.hg19_VCAN","8.hg19_LTB","8.hg19_TCF7","8.hg19_IL7R","8.hg19_LDHB","8.hg19_CD27","8.hg19_LEF1","8.hg19_NOSIP","8.hg19_SELL","8.hg19_NPM1","8.hg19_CAMK4","9.hg19_CCR7","9.hg19_ACTN1","9.hg19_NELL2","9.hg19_PASK","9.hg19_MAL","9.hg19_CD8B","9.hg19_LEF1","9.hg19_PIK3IP1","9.hg19_LDLRAP1","9.hg19_PCED1B"],["2668"," 850","2126","2159","3396","3393","2878","3109","3983","3478","3603","2742","3449","3498","2047","4272","2804","2748","3121","3143","3340","3350","4081","4764","5470","5452","5686","5693","5469","5538","2588","3314","3461","3771","4262","4495","4462","4328","4506","4372","4925","4728","4462","4388","5128","7448","7474","6322","6835","6392","2260","3760","4760","4706","4978","4723","4865","6370","5498","6103","2400","2406","2409","2600"," 661","2808","3200","2825","3400","3400","3843","5576","5575","5776","8738","8339","7978","8482","7152","9506","2449","2815","2907","3163","3115","2175","2384","2787","3260","3524"],["7.8e-24","2.5e-23","1.1e-11","1.2e-10","1.6e-10","3.2e-10","9.2e-09","2.5e-07","7.5e-06","7.6e-05","7.0e-52","3.7e-37","4.7e-34","1.4e-33","1.6e-33","6.8e-31","1.4e-27","2.3e-27","8.4e-26","3.1e-24","3.1e-16","8.1e-11","3.7e-09","9.7e-09","1.6e-05","3.1e-05","2.1e-04","2.7e-04","2.7e-04","3.7e-04","2.6e-09","1.1e-05","2.6e-05","1.8e-04","7.1e-04","1.5e-03","1.6e-03","2.0e-03","2.1e-03","2.2e-03","2.3e-42","1.6e-37","1.1e-22","2.1e-22","5.5e-19","8.9e-14","7.2e-13","9.0e-13","1.7e-12","1.9e-12","1.4e-22","1.2e-15","1.5e-13","2.9e-13","2.3e-12","1.5e-11","1.8e-11","5.1e-11","1.7e-10","1.5e-09","4.5e-64","1.1e-61","1.3e-61","1.9e-61","9.1e-60","2.2e-56","1.3e-53","4.7e-52","4.8e-51","4.8e-51","1.1e-33","2.2e-30","8.0e-24","5.5e-21","6.7e-20","5.2e-19","2.4e-15","2.9e-15","6.4e-15","1.7e-13","3.0e-21","3.7e-19","8.1e-19","1.6e-18","1.4e-13","2.8e-13","5.0e-13","5.0e-11","1.5e-10","2.7e-10"],["7.8e-24","2.5e-23","1.1e-11","1.2e-10","1.6e-10","3.2e-10","9.2e-09","2.5e-07","7.5e-06","7.6e-05","7.0e-52","3.7e-37","4.7e-34","1.4e-33","1.6e-33","6.8e-31","1.4e-27","2.3e-27","8.4e-26","3.1e-24","3.1e-16","8.1e-11","3.7e-09","9.7e-09","1.6e-05","3.1e-05","2.1e-04","2.7e-04","2.7e-04","3.7e-04","2.6e-09","1.1e-05","2.6e-05","1.8e-04","7.1e-04","1.5e-03","1.6e-03","2.0e-03","2.1e-03","2.2e-03","2.3e-42","1.6e-37","1.1e-22","2.1e-22","5.5e-19","8.9e-14","7.2e-13","9.0e-13","1.7e-12","1.9e-12","1.4e-22","1.2e-15","1.5e-13","2.9e-13","2.3e-12","1.5e-11","1.8e-11","5.1e-11","1.7e-10","1.5e-09","4.5e-64","1.1e-61","1.3e-61","1.9e-61","9.1e-60","2.2e-56","1.3e-53","4.7e-52","4.8e-51","4.8e-51","1.1e-33","2.2e-30","8.0e-24","5.5e-21","6.7e-20","5.2e-19","2.4e-15","2.9e-15","6.4e-15","1.7e-13","3.0e-21","3.7e-19","8.1e-19","1.6e-18","1.4e-13","2.8e-13","5.0e-13","5.0e-11","1.5e-10","2.7e-10"],["0.1088","0.4686","0.5104","0.9285","0.1657","0.2075","0.4686","0.4013","0.2302","1.7839","0.0184","0.1293","0.0921","0.1812","0.2495","0.0941","0.2766","0.4647","0.2215","0.4411","0.3411","1.9895","0.8091","0.4066","0.4660","0.5375","0.5282","0.6042","1.1880","1.3497","0.9398","2.3237","1.2814","1.7858","0.4562","0.3672","0.4132","0.6647","0.4315","0.6902","0.1020","0.1152","0.6229","0.6790","0.7474","0.3320","0.2744","1.3792","0.6761","1.8741","2.3051","1.9122","0.6290","0.9245","0.8180","1.6226","2.0948","0.2391","0.7842","0.4878","0.0000","0.0016","0.0027","0.0000","0.0529","0.0020","0.0000","0.0079","0.0000","0.0000","0.5670","0.2327","0.7453","0.7186","0.1238","0.1807","0.3746","0.3341","1.2434","0.1818","0.0890","0.0690","0.0740","0.0513","0.1080","0.3614","0.2537","0.2096","0.1353","0.0991"],["0.92","2.80","1.69","2.53","0.81","0.94","1.28","1.07","0.69","2.67","0.91","1.22","1.42","1.71","1.81","0.96","1.78","2.11","1.34","2.02","1.39","3.81","1.82","1.33","1.07","1.10","0.99","1.09","1.63","1.94","2.33","3.31","1.99","2.62","0.83","0.79","0.76","1.09","0.82","1.18","1.66","1.15","2.01","2.06","2.07","1.00","0.94","2.64","1.56","3.64","5.22","3.88","1.61","2.21","1.96","2.59","3.93","0.78","1.74","1.13","1.83","2.74","1.14","0.81","2.84","2.53","0.74","2.81","1.04","1.50","2.38","1.20","2.20","1.79","0.61","0.82","1.04","1.04","2.07","0.62","0.82","0.65","0.56","0.51","0.60","1.57","1.09","0.86","0.66","0.46"],["0.0565","0.2285","0.3489","0.4251","0.1278","0.1351","0.2998","0.3464","0.1843","0.7617","0.0130","0.0935","0.0831","0.0779","0.1766","0.0597","0.1896","0.1818","0.1844","0.2000","0.2127","0.6101","0.4127","0.2430","0.3089","0.3570","0.4101","0.4329","0.8076","0.7544","0.4236","0.8941","0.7438","0.7660","0.3793","0.3153","0.3350","0.4828","0.3374","0.4581","0.0438","0.0767","0.3397","0.3699","0.3726","0.1945","0.2356","0.5589","0.3753","0.5781","0.5748","0.5958","0.3753","0.4094","0.3990","0.8294","0.6010","0.1811","0.3990","0.2992","0.0000","0.0025","0.0025","0.0000","0.0450","0.0025","0.0000","0.0075","0.0000","0.0000","0.2951","0.1719","0.3352","0.5100","0.1032","0.1461","0.3095","0.2264","0.7221","0.1633","0.0784","0.0686","0.0637","0.0515","0.0931","0.2304","0.1985","0.1838","0.1275","0.0980"],["0.63","0.97","0.90","0.93","0.57","0.57","0.83","0.77","0.53","1.00","0.65","0.81","0.69","0.73","0.88","0.63","0.81","0.94","0.79","0.87","0.79","1.00","0.90","0.62","0.60","0.67","0.64","0.69","0.93","0.93","0.94","1.00","0.97","0.94","0.65","0.55","0.61","0.74","0.58","0.68","0.67","0.71","0.90","0.92","0.92","0.64","0.60","0.99","0.82","1.00","1.00","1.00","0.86","0.93","0.89","0.98","1.00","0.59","0.88","0.73","0.68","0.68","0.68","0.65","0.92","0.62","0.57","0.62","0.54","0.54","0.93","0.78","0.95","0.86","0.53","0.59","0.72","0.66","0.95","0.55","0.66","0.59","0.59","0.52","0.59","0.79","0.76","0.72","0.55","0.52"],["0.81","2.33","1.18","1.61","0.65","0.74","0.81","0.67","0.46","0.89","0.89","1.09","1.33","1.52","1.56","0.87","1.51","1.65","1.12","1.58","1.05","1.82","1.01","0.92","0.61","0.56","0.47","0.49","0.45","0.59","1.39","0.98","0.71","0.83","0.37","0.42","0.35","0.42","0.39","0.49","1.56","1.03","1.38","1.38","1.32","0.67","0.67","1.26","0.89","1.76","2.92","1.97","0.98","1.29","1.15","0.96","1.83","0.54","0.95","0.64","1.83","2.74","1.14","0.81","2.79","2.52","0.74","2.81","1.04","1.50","1.81","0.97","1.46","1.07","0.49","0.64","0.67","0.71","0.83","0.44","0.73","0.58","0.49","0.46","0.49","1.21","0.84","0.65","0.52","0.36"],["0.58","0.74","0.55","0.51","0.44","0.43","0.53","0.42","0.35","0.24","0.64","0.71","0.61","0.65","0.71","0.57","0.62","0.76","0.60","0.67","0.57","0.39","0.49","0.38","0.29","0.31","0.23","0.26","0.12","0.17","0.51","0.11","0.22","0.17","0.27","0.23","0.28","0.26","0.24","0.22","0.62","0.63","0.56","0.55","0.54","0.44","0.36","0.43","0.44","0.42","0.43","0.40","0.48","0.52","0.49","0.15","0.40","0.41","0.48","0.43","0.68","0.67","0.67","0.65","0.87","0.62","0.57","0.61","0.54","0.54","0.64","0.61","0.62","0.35","0.43","0.44","0.41","0.43","0.23","0.38","0.58","0.52","0.52","0.47","0.49","0.56","0.56","0.54","0.42","0.42"],["hg19_GZMK","hg19_KLRB1","hg19_DUSP2","hg19_IL7R","hg19_NCR3","hg19_CEBPD","hg19_KLRG1","hg19_CD69","hg19_LINC01871","hg19_IL32","hg19_SH2D1B","hg19_KLRF1","hg19_CLIC3","hg19_FCER1G","hg19_MYOM2","hg19_TRDC","hg19_SPON2","hg19_TYROBP","hg19_IL2RB","hg19_KLRB1","hg19_CD8B","hg19_CCL5","hg19_CD8A","hg19_CMC1","hg19_KLRG1","hg19_DUSP2","hg19_FYN","hg19_GZMM","hg19_SRP14","hg19_HCST","hg19_IL7R","hg19_S100A4","hg19_S100A10","hg19_IL32","hg19_FAM107B","hg19_EML4","hg19_CD6","hg19_CD2","hg19_RORA","hg19_ITGB1","hg19_TRBV20-1","hg19_ZNF683","hg19_FGFBP2","hg19_CD8A","hg19_GZMH","hg19_CD8B","hg19_CYTOR","hg19_CST7","hg19_KLRD1","hg19_CCL5","hg19_GNLY","hg19_CCL5","hg19_CTSW","hg19_GZMB","hg19_GZMH","hg19_IFITM2","hg19_NKG7","hg19_PRSS23","hg19_CD8A","hg19_HOPX","hg19_FCN1","hg19_LYZ","hg19_SPI1","hg19_CYBB","hg19_HLA-DRA","hg19_S100A9","hg19_CD68","hg19_S100A8","hg19_CSTA","hg19_VCAN","hg19_LTB","hg19_TCF7","hg19_IL7R","hg19_LDHB","hg19_CD27","hg19_LEF1","hg19_NOSIP","hg19_SELL","hg19_NPM1","hg19_CAMK4","hg19_CCR7","hg19_ACTN1","hg19_NELL2","hg19_PASK","hg19_MAL","hg19_CD8B","hg19_LEF1","hg19_PIK3IP1","hg19_LDLRAP1","hg19_PCED1B"],["1","1","1","1","1","1","1","1","1","1","2","2","2","2","2","2","2","2","2","2","3","3","3","3","3","3","3","3","3","3","4","4","4","4","4","4","4","4","4","4","5","5","5","5","5","5","5","5","5","5","6","6","6","6","6","6","6","6","6","6","7","7","7","7","7","7","7","7","7","7","8","8","8","8","8","8","8","8","8","8","9","9","9","9","9","9","9","9","9","9"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stats.W<\/th>\n      <th>pval<\/th>\n      <th>p.adjust<\/th>\n      <th>meanExprs.1<\/th>\n      <th>meanExprs.2<\/th>\n      <th>meanPct.1<\/th>\n      <th>meanPct.2<\/th>\n      <th>meanDiff<\/th>\n      <th>pctDiff<\/th>\n      <th>name<\/th>\n      <th>group<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="for-adt-count" class="section level3">
<h3 class="hasAnchor">
<a href="#for-adt-count" class="anchor"></a>For ADT count</h3>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEgenes.html">DEgenes</a></span>(<span class="no">sce_citeseq</span>,
                       <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
                       <span class="kw">group</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>$<span class="no">SNF_W_louvain</span>,
                       <span class="kw">return_all</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                       <span class="kw">exprs_pct</span> <span class="kw">=</span> <span class="fl">0.5</span>)
<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/selectDEgenes.html">selectDEgenes</a></span>(<span class="no">sce_citeseq</span>,
                             <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/format.html">format</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="no">rbind</span>, <span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"DE_res_ADT_filter"</span>]]),
                 <span class="kw">digits</span> <span class="kw">=</span> <span class="fl">2</span>))</pre></body></html></div>
<div id="htmlwidget-f3aa05f664768818f55c" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-f3aa05f664768818f55c">{"x":{"filter":"none","data":[["1.CD26 (Adenosine)","1.CD28","2","7.LAMP1","7.CD45RO","7.CD45","7.B220 (CD45R)","7.CD14","7.B7-H1 (PD-L1)","7.Siglec-8","7.PD-L1 (CD274)","7.TCRg","7.EpCAM (CD326)","8.CD28","8.CD26 (Adenosine)","8.CD3","9"],[" 2830"," 3124"," 3146"," 1652"," 1660"," 2274"," 2366"," 3156"," 2859"," 2594"," 2604"," 2774"," 2860"," 4106","10616","10628"," 4549"],["5.0e-08","7.3e-06","4.9e-17","2.0e-15","3.0e-15","2.9e-13","2.0e-12","1.2e-11","1.7e-11","2.0e-11","4.1e-11","4.2e-11","8.1e-11","1.3e-26","6.5e-07","7.4e-06","3.7e-02"],["5.0e-08","7.3e-06","4.9e-17","2.0e-15","3.0e-15","2.9e-13","2.0e-12","1.2e-11","1.7e-11","2.0e-11","4.1e-11","4.2e-11","8.1e-11","1.3e-26","6.5e-07","7.4e-06","3.7e-02"],["0.46","1.48","0.65","0.92","1.13","0.57","0.79","0.27","0.41","0.78","1.13","0.56","0.54","1.29","0.45","1.43","1.51"],["1.59","2.28","1.74","2.55","2.59","1.46","1.71","1.59","1.32","1.80","2.16","1.44","1.51","2.54","0.87","1.83","1.87"],["0.41","0.85","0.54","0.77","0.85","0.58","0.75","0.33","0.46","0.72","0.87","0.57","0.56","0.83","0.38","0.85","0.85"],["0.73","0.97","0.94","1.00","1.00","0.95","0.97","0.70","0.84","0.95","0.97","0.86","0.89","0.99","0.64","0.97","1.00"],["1.14","0.79","1.09","1.63","1.46","0.89","0.93","1.32","0.91","1.03","1.04","0.88","0.96","1.26","0.42","0.40","0.36"],["0.32","0.11","0.40","0.23","0.15","0.37","0.22","0.38","0.38","0.22","0.10","0.30","0.33","0.16","0.26","0.11","0.15"],["CD26 (Adenosine)","CD28","CD366 (tim3)","LAMP1","CD45RO","CD45","B220 (CD45R)","CD14","B7-H1 (PD-L1)","Siglec-8","PD-L1 (CD274)","TCRg","EpCAM (CD326)","CD28","CD26 (Adenosine)","CD3","CD28"],["1","1","2","7","7","7","7","7","7","7","7","7","7","8","8","8","9"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stats.W<\/th>\n      <th>pval<\/th>\n      <th>p.adjust<\/th>\n      <th>meanExprs.1<\/th>\n      <th>meanExprs.2<\/th>\n      <th>meanPct.1<\/th>\n      <th>meanPct.2<\/th>\n      <th>meanDiff<\/th>\n      <th>pctDiff<\/th>\n      <th>name<\/th>\n      <th>group<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="visualising-de-results" class="section level3">
<h3 class="hasAnchor">
<a href="#visualising-de-results" class="anchor"></a>Visualising DE Results</h3>
<p>The DE genes can be visualised with the <code>DEbubblePlot</code> and <code>DEcomparisonPlot</code>. In each case, the gene names must first be extracted from the DE result objects.</p>
<div id="circlepackplot" class="section level4">
<h4 class="hasAnchor">
<a href="#circlepackplot" class="anchor"></a>circlepackPlot</h4>
<p>The <code>circlepackPlot</code> takes a list of all DE genes from RNA and ADT DE analysis and will plot only the top most significant DE genes to plot.</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="no">rna_DEgenes</span> <span class="kw">&lt;-</span> <span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"DE_res_RNA_filter"</span>]]
<span class="no">adt_DEgenes</span> <span class="kw">&lt;-</span> <span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"DE_res_ADT_filter"</span>]]
<span class="no">rna_DEgenes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">rna_DEgenes</span>, <span class="kw">function</span>(<span class="no">x</span>){
  <span class="no">x</span>$<span class="no">name</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"hg19_"</span>, <span class="st">""</span>, <span class="no">x</span>$<span class="no">name</span>)
  <span class="no">x</span>})
<span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEbubblePlot.html">DEbubblePlot</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">RNA</span> <span class="kw">=</span> <span class="no">rna_DEgenes</span>, <span class="kw">ADT</span> <span class="kw">=</span> <span class="no">adt_DEgenes</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-28-1.png" width="960"></p>
</div>
<div id="decomparisonplot" class="section level4">
<h4 class="hasAnchor">
<a href="#decomparisonplot" class="anchor"></a>DEcomparisonPlot</h4>
<p>For the <code>DEcomparisonPlot</code>, as well as a list containing the DE genes for RNA and ADT, a <code>feature_list</code> specifying the genes and proteins of interest is required.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">rna_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"hg19_CD4"</span>,
              <span class="st">"hg19_CD8A"</span>,
              <span class="st">"hg19_HLA-DRB1"</span>,
              <span class="st">"hg19_ITGAX"</span>,
              <span class="st">"hg19_NCAM1"</span>,
              <span class="st">"hg19_CD27"</span>,
              <span class="st">"hg19_CD19"</span>)
<span class="no">adt_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"MHCII (HLA-DR)"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD56"</span>, <span class="st">"CD27"</span>, <span class="st">"CD19"</span>)
<span class="no">rna_DEgenes_all</span> <span class="kw">&lt;-</span> <span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"DE_res_RNA"</span>]]
<span class="no">adt_DEgenes_all</span> <span class="kw">&lt;-</span> <span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"DE_res_ADT"</span>]]
<span class="no">feature_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">RNA</span> <span class="kw">=</span> <span class="no">rna_list</span>, <span class="kw">ADT</span> <span class="kw">=</span> <span class="no">adt_list</span>)
<span class="no">de_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">RNA</span> <span class="kw">=</span> <span class="no">rna_DEgenes_all</span>, <span class="kw">ADT</span> <span class="kw">=</span> <span class="no">adt_DEgenes_all</span>)
<span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEcomparisonPlot.html">DEcomparisonPlot</a></span>(<span class="kw">de_list</span> <span class="kw">=</span> <span class="no">de_list</span>,
                 <span class="kw">feature_list</span> <span class="kw">=</span> <span class="no">feature_list</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-29-1.png" width="768"></p>
</div>
</div>
</div>
<div id="adt-importance-evaluation" class="section level2">
<h2 class="hasAnchor">
<a href="#adt-importance-evaluation" class="anchor"></a>ADT Importance Evaluation</h2>
<p>An important evaluation in CITE-seq data analysis is to assess the quality of each ADT and to evaluate the contribution of ADTs towards clustering outcome. CiteFuse calculates the relative importance of ADT towards clustering outcome by using a random forest model. The higher the score of an ADT, the greater its importance towards the final clustering outcome.</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">2020</span>)
<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/importanceADT.html">importanceADT</a></span>(<span class="no">sce_citeseq</span>,
                             <span class="kw">group</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>$<span class="no">SNF_W_louvain</span>,
                             <span class="kw">subsample</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visImportance.html">visImportance</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"boxplot"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-30-1.png" width="768"></p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visImportance.html">visImportance</a></span>(<span class="no">sce_citeseq</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"heatmap"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-30-2.png" width="768"></p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"importanceADT"</span>]], <span class="kw">decreasing</span> <span class="kw">=</span> <span class="fl">TRUE</span>)[<span class="fl">1</span>:<span class="fl">20</span>]
<span class="co">#&gt;               CD4              CD27               CD8               CD5 </span>
<span class="co">#&gt;         30.036201         29.449213         28.557098         25.486926 </span>
<span class="co">#&gt;               CD7      PECAM (CD31)             CD11b               CD2 </span>
<span class="co">#&gt;         16.493928         12.961868         11.988086         11.766563 </span>
<span class="co">#&gt;              CD28              CD44    MHCII (HLA-DR) IL7Ralpha (CD127) </span>
<span class="co">#&gt;          9.947612          9.788229          9.651328          8.516038 </span>
<span class="co">#&gt;         HLA-A,B,C             CD62L      CD366 (tim3)            CD45RA </span>
<span class="co">#&gt;          7.126189          5.710455          5.670500          5.659654 </span>
<span class="co">#&gt;               CD3             CD11c              CD56      PD-1 (CD279) </span>
<span class="co">#&gt;          5.481791          5.202686          5.024104          4.637673</span></pre></body></html></div>
<p>The importance scores can be visualised in a boxplot and heatmap. Our evaluation of ADT importance show that unsurprisingly CD4 and CD8 are the top two discriminating proteins in PBMCs.</p>
<p>Let us try clustering with only ADTs with a score greater than 5.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="no">subset_adt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"importanceADT"</span>]] <span class="kw">&gt;</span> <span class="fl">5</span>))
<span class="no">subset_adt</span>
<span class="co">#&gt;  [1] "CD11b"             "CD11c"             "CD2"              </span>
<span class="co">#&gt;  [4] "CD27"              "CD28"              "CD3"              </span>
<span class="co">#&gt;  [7] "CD366 (tim3)"      "CD4"               "CD44"             </span>
<span class="co">#&gt; [10] "CD45RA"            "CD5"               "CD56"             </span>
<span class="co">#&gt; [13] "CD62L"             "CD7"               "CD8"              </span>
<span class="co">#&gt; [16] "HLA-A,B,C"         "IL7Ralpha (CD127)" "MHCII (HLA-DR)"   </span>
<span class="co">#&gt; [19] "PECAM (CD31)"</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/CiteFuse.html">CiteFuse</a></span>(<span class="no">sce_citeseq</span>,
                                    <span class="kw">ADT_subset</span> <span class="kw">=</span> <span class="no">subset_adt</span>,
                                    <span class="kw">metadata_names</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"W_SNF_adtSubset1"</span>,
                                                       <span class="st">"W_ADT_adtSubset1"</span>,
                                                       <span class="st">"W_RNA"</span>)))
<span class="co">#&gt; Calculating affinity matrix </span>
<span class="co">#&gt; Performing SNF</span>
<span class="co">#&gt;    user  system elapsed </span>
<span class="co">#&gt;   4.366   0.049   4.419</span>
<span class="no">SNF_W_clust_adtSubset1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html">spectralClustering</a></span>(<span class="fu">metadata</span>(<span class="no">sce_citeseq</span>)<span class="kw">[[</span><span class="st">"W_SNF_adtSubset1"</span>]], <span class="kw">K</span> <span class="kw">=</span> <span class="fl">5</span>)
<span class="co">#&gt; Computing Spectral Clustering</span>
<span class="no">sce_citeseq</span>$<span class="no">SNF_W_clust_adtSubset1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">SNF_W_clust_adtSubset1</span>$<span class="no">labels</span>)

<span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html">adjustedRandIndex</a></span>(<span class="no">sce_citeseq</span>$<span class="no">SNF_W_clust_adtSubset1</span>, <span class="no">sce_citeseq</span>$<span class="no">SNF_W_clust</span>)
<span class="co">#&gt; [1] 0.9332215</span></pre></body></html></div>
<p>When we compare between the two clustering outcomes, we find that the adjusted rand index is approximately 0.93, where a value of 1 denotes complete concordance.</p>
</div>
<div id="rna-ligand---adt-receptor-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#rna-ligand---adt-receptor-analysis" class="anchor"></a>RNA Ligand - ADT Receptor Analysis</h2>
<p>With the advent of CITE-seq, we can now predict ligand-receptor interactions by using cell surface protein expression. CiteFuse implements a <code>ligandReceptorTest</code> to find ligand receptor interactions between sender and receiver cells. Importantly, the ADT count is used to predict receptor expression within receiver cells. Note that the setting <code>altExp_name = "RNA"</code> would enable users to predict ligand-receptor interaction from RNA expression only.</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"lr_pair_subset"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"CiteFuse"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">lr_pair_subset</span>)
<span class="co">#&gt;      [,1]          [,2]   </span>
<span class="co">#&gt; [1,] "hg19_IL17RA" "CD45" </span>
<span class="co">#&gt; [2,] "hg19_FAS"    "CD11b"</span>
<span class="co">#&gt; [3,] "hg19_GZMK"   "CD62L"</span>
<span class="co">#&gt; [4,] "hg19_CD40LG" "CD11b"</span>
<span class="co">#&gt; [5,] "hg19_FLT3LG" "CD62L"</span>
<span class="co">#&gt; [6,] "hg19_GZMA"   "CD19"</span>
<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html">normaliseExprs</a></span>(<span class="kw">sce</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>,
                              <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
                              <span class="kw">transform</span> <span class="kw">=</span> <span class="st">"zi_minMax"</span>)
<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html">normaliseExprs</a></span>(<span class="kw">sce</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>,
                              <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"none"</span>,
                              <span class="kw">exprs_value</span> <span class="kw">=</span> <span class="st">"logcounts"</span>,
                              <span class="kw">transform</span> <span class="kw">=</span> <span class="st">"minMax"</span>)
<span class="no">sce_citeseq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/ligandReceptorTest.html">ligandReceptorTest</a></span>(<span class="kw">sce</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>,
                                  <span class="kw">ligandReceptor_list</span> <span class="kw">=</span> <span class="no">lr_pair_subset</span>,
                                  <span class="kw">cluster</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>$<span class="no">SNF_W_louvain</span>,
                                  <span class="kw">RNA_exprs_value</span> <span class="kw">=</span> <span class="st">"minMax"</span>,
                                  <span class="kw">use_alt_exp</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                                  <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
                                  <span class="kw">altExp_exprs_value</span> <span class="kw">=</span> <span class="st">"zi_minMax"</span>,
                                  <span class="kw">num_permute</span> <span class="kw">=</span> <span class="fl">1000</span>)
<span class="co">#&gt; 100 ......200 ......300 ......400 ......500 ......600 ......700 ......800 ......900 ......1000 ......</span></pre></body></html></div>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html">visLigandReceptor</a></span>(<span class="no">sce_citeseq</span>,
                  <span class="kw">type</span> <span class="kw">=</span> <span class="st">"pval_heatmap"</span>,
                  <span class="kw">receptor_type</span> <span class="kw">=</span> <span class="st">"ADT"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-33-1.png" width="768"></p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html">visLigandReceptor</a></span>(<span class="no">sce_citeseq</span>,
                  <span class="kw">type</span> <span class="kw">=</span> <span class="st">"pval_dotplot"</span>,
                  <span class="kw">receptor_type</span> <span class="kw">=</span> <span class="st">"ADT"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-34-1.png" width="576"></p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html">visLigandReceptor</a></span>(<span class="no">sce_citeseq</span>,
                  <span class="kw">type</span> <span class="kw">=</span> <span class="st">"group_network"</span>,
                  <span class="kw">receptor_type</span> <span class="kw">=</span> <span class="st">"ADT"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-35-1.png" width="768"></p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html">visLigandReceptor</a></span>(<span class="no">sce_citeseq</span>,
                  <span class="kw">type</span> <span class="kw">=</span> <span class="st">"group_heatmap"</span>,
                  <span class="kw">receptor_type</span> <span class="kw">=</span> <span class="st">"ADT"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-36-1.png" width="768"></p>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html">visLigandReceptor</a></span>(<span class="no">sce_citeseq</span>,
                  <span class="kw">type</span> <span class="kw">=</span> <span class="st">"lr_network"</span>,
                  <span class="kw">receptor_type</span> <span class="kw">=</span> <span class="st">"ADT"</span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-37-1.png" width="768"></p>
</div>
</div>
<div id="extension" class="section level1">
<h1 class="hasAnchor">
<a href="#extension" class="anchor"></a>Extension</h1>
<div id="between-sample-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#between-sample-analysis" class="anchor"></a>Between-sample analysis</h2>
<p>Lastly, we will jointly analyse the current PBMC CITE-seq data, taken from healthy human donors, and another subset of CITE-seq data from patients with cutaneous T-cell lymphoma (CTCL), again from Mimitou et al. (2019). The data <code>sce_ctcl_subset</code> provided in our <code>CiteFuse</code> package already contains the clustering information.</p>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"sce_ctcl_subset"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"CiteFuse"</span>)</pre></body></html></div>
<p>To visualise and compare gene or protein expression data, we can use <code>visualiseExprsList</code> function.</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprsList.html">visualiseExprsList</a></span>(<span class="kw">sce_list</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">control</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>,
                                   <span class="kw">ctcl</span> <span class="kw">=</span> <span class="no">sce_ctcl_subset</span>),
                   <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"boxplot"</span>,
                   <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"none"</span>,
                   <span class="kw">exprs_value</span> <span class="kw">=</span> <span class="st">"logcounts"</span>,
                   <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"hg19_S100A10"</span>, <span class="st">"hg19_CD8A"</span>),
                   <span class="kw">group_by</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"SNF_W_louvain"</span>, <span class="st">"SNF_W_louvain"</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-39-1.png" width="700"></p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprsList.html">visualiseExprsList</a></span>(<span class="kw">sce_list</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">control</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>,
                                   <span class="kw">ctcl</span> <span class="kw">=</span> <span class="no">sce_ctcl_subset</span>),
                   <span class="kw">plot</span> <span class="kw">=</span> <span class="st">"boxplot"</span>,
                   <span class="kw">altExp_name</span> <span class="kw">=</span> <span class="st">"ADT"</span>,
                   <span class="kw">feature_subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CD19"</span>, <span class="st">"CD8"</span>),
                   <span class="kw">group_by</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"SNF_W_louvain"</span>, <span class="st">"SNF_W_louvain"</span>))</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-39-2.png" width="700"></p>
<p>We can then perform differential expression analysis of the RNA expression level across the two clusters that have high CD19 expression in ADT.</p>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="no">de_res</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEgenesCross.html">DEgenesCross</a></span>(<span class="kw">sce_list</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">control</span> <span class="kw">=</span> <span class="no">sce_citeseq</span>,
                                       <span class="kw">ctcl</span> <span class="kw">=</span> <span class="no">sce_ctcl_subset</span>),
                       <span class="kw">colData_name</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"SNF_W_louvain"</span>, <span class="st">"SNF_W_louvain"</span>),
                       <span class="kw">group_to_test</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"2"</span>, <span class="st">"6"</span>))
<span class="no">de_res_filter</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/selectDEgenes.html">selectDEgenes</a></span>(<span class="kw">de_res</span> <span class="kw">=</span> <span class="no">de_res</span>)
<span class="no">de_res_filter</span>
<span class="co">#&gt; $control</span>
<span class="co">#&gt;             stats.W         pval     p.adjust meanExprs.1 meanExprs.2</span>
<span class="co">#&gt; hg19_GNLY       0.0 2.301890e-18 2.301890e-18  0.07220401    5.193820</span>
<span class="co">#&gt; hg19_EFHD2     67.0 9.718826e-17 9.718826e-17  0.02064028    1.810989</span>
<span class="co">#&gt; hg19_CST7      41.0 1.063099e-16 1.063099e-16  0.22512314    2.884901</span>
<span class="co">#&gt; hg19_NKG7      38.0 2.377913e-16 2.377913e-16  0.51066894    4.283951</span>
<span class="co">#&gt; hg19_GZMB      78.0 4.247635e-16 4.247635e-16  0.11667246    2.646837</span>
<span class="co">#&gt; hg19_PRF1     133.5 5.541530e-15 5.541530e-15  0.06473009    1.974489</span>
<span class="co">#&gt; hg19_ITGB2    110.0 2.714691e-14 2.714691e-14  0.44185664    2.227816</span>
<span class="co">#&gt; hg19_CYBA     101.0 2.944019e-14 2.944019e-14  1.22818319    2.575690</span>
<span class="co">#&gt; hg19_FGFBP2   184.0 4.298063e-14 4.298063e-14  0.03101677    1.769919</span>
<span class="co">#&gt; hg19_IFITM2   126.0 1.227095e-13 1.227095e-13  1.47843667    3.018231</span>
<span class="co">#&gt;              meanPct.1 meanPct.2 meanDiff   pctDiff        name   group</span>
<span class="co">#&gt; hg19_GNLY   0.02325581 1.0000000 5.121616 0.9767442   hg19_GNLY control</span>
<span class="co">#&gt; hg19_EFHD2  0.02325581 0.9423077 1.790349 0.9190519  hg19_EFHD2 control</span>
<span class="co">#&gt; hg19_CST7   0.13953488 1.0000000 2.659777 0.8604651   hg19_CST7 control</span>
<span class="co">#&gt; hg19_NKG7   0.30232558 1.0000000 3.773283 0.6976744   hg19_NKG7 control</span>
<span class="co">#&gt; hg19_GZMB   0.06976744 0.9615385 2.530164 0.8917710   hg19_GZMB control</span>
<span class="co">#&gt; hg19_PRF1   0.04651163 0.9038462 1.909759 0.8573345   hg19_PRF1 control</span>
<span class="co">#&gt; hg19_ITGB2  0.44186047 0.9615385 1.785960 0.5196780  hg19_ITGB2 control</span>
<span class="co">#&gt; hg19_CYBA   0.88372093 1.0000000 1.347507 0.1162791   hg19_CYBA control</span>
<span class="co">#&gt; hg19_FGFBP2 0.02325581 0.8461538 1.738903 0.8228980 hg19_FGFBP2 control</span>
<span class="co">#&gt; hg19_IFITM2 0.88372093 1.0000000 1.539794 0.1162791 hg19_IFITM2 control</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ctcl</span>
<span class="co">#&gt;               stats.W         pval     p.adjust meanExprs.1 meanExprs.2</span>
<span class="co">#&gt; hg19_RPS26          0 6.431902e-17 6.431902e-17  1.48598738    5.072471</span>
<span class="co">#&gt; hg19_CD8B         160 6.208782e-16 6.208782e-16  0.01600345    1.618837</span>
<span class="co">#&gt; hg19_LTB          157 1.064038e-14 1.064038e-14  0.16359138    1.978730</span>
<span class="co">#&gt; hg19_IL7R         247 2.905360e-13 2.905360e-13  0.08949349    1.651591</span>
<span class="co">#&gt; hg19_RPL36A       181 2.328401e-12 2.328401e-12  1.27615800    2.846750</span>
<span class="co">#&gt; hg19_CD3D         300 2.399810e-12 2.399810e-12  0.08958797    1.222056</span>
<span class="co">#&gt; hg19_RPL7         280 3.740249e-10 3.740249e-10  1.47545170    2.618403</span>
<span class="co">#&gt; hg19_LEPROTL1     314 4.234225e-10 4.234225e-10  0.34891667    1.300745</span>
<span class="co">#&gt; hg19_NPM1         291 5.346323e-10 5.346323e-10  1.00758312    2.202268</span>
<span class="co">#&gt; hg19_TMEM123      368 5.569121e-10 5.569121e-10  0.15371329    1.012034</span>
<span class="co">#&gt;                meanPct.1 meanPct.2  meanDiff   pctDiff          name group</span>
<span class="co">#&gt; hg19_RPS26    0.86538462 1.0000000 3.5864831 0.1346154    hg19_RPS26  ctcl</span>
<span class="co">#&gt; hg19_CD8B     0.01923077 0.8604651 1.6028340 0.8412343     hg19_CD8B  ctcl</span>
<span class="co">#&gt; hg19_LTB      0.13461538 0.9069767 1.8151388 0.7723614      hg19_LTB  ctcl</span>
<span class="co">#&gt; hg19_IL7R     0.07692308 0.8139535 1.5620970 0.7370304     hg19_IL7R  ctcl</span>
<span class="co">#&gt; hg19_RPL36A   0.73076923 1.0000000 1.5705921 0.2692308   hg19_RPL36A  ctcl</span>
<span class="co">#&gt; hg19_CD3D     0.03846154 0.7906977 1.1324682 0.7522361     hg19_CD3D  ctcl</span>
<span class="co">#&gt; hg19_RPL7     0.82692308 1.0000000 1.1429510 0.1730769     hg19_RPL7  ctcl</span>
<span class="co">#&gt; hg19_LEPROTL1 0.30769231 0.9069767 0.9518288 0.5992844 hg19_LEPROTL1  ctcl</span>
<span class="co">#&gt; hg19_NPM1     0.65384615 0.9534884 1.1946849 0.2996422     hg19_NPM1  ctcl</span>
<span class="co">#&gt; hg19_TMEM123  0.15384615 0.7674419 0.8583209 0.6135957  hg19_TMEM123  ctcl</span></pre></body></html></div>
</div>
<div id="gene---adt-network" class="section level2">
<h2 class="hasAnchor">
<a href="#gene---adt-network" class="anchor"></a>Gene - ADT network</h2>
<p>The <code>geneADTnetwork</code> function plots an interaction network between genes identified from the DE analysis. The nodes denote proteins and RNA whilst the edges denote positive and negative correlation in expression.</p>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="no">RNA_feature_subset</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">rna_DEgenes_all</span>, <span class="st">"[["</span>, <span class="st">"name"</span>))))
<span class="no">ADT_feature_subset</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">adt_DEgenes_all</span>, <span class="st">"[["</span>, <span class="st">"name"</span>))))
<span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/geneADTnetwork.html">geneADTnetwork</a></span>(<span class="no">sce_citeseq</span>,
               <span class="kw">RNA_feature_subset</span> <span class="kw">=</span> <span class="no">RNA_feature_subset</span>,
               <span class="kw">ADT_feature_subset</span> <span class="kw">=</span> <span class="no">ADT_feature_subset</span>,
               <span class="kw">cor_method</span> <span class="kw">=</span> <span class="st">"pearson"</span>,
               <span class="kw">network_layout</span> <span class="kw">=</span> <span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/igraph/man/layout_with_fr.html">layout_with_fr</a></span>)</pre></body></html></div>
<p><img src="CiteFuse_BioCAsia_workshop_files/figure-html/unnamed-chunk-41-1.png" width="768"></p>
<pre><code>#&gt; IGRAPH 499e405 UN-B 72 134 -- 
#&gt; + attr: name (v/c), label (v/c), class (v/c), type (v/l), shape (v/c),
#&gt; | color (v/c), size (v/n), label.cex (v/n), label.color (v/c), value
#&gt; | (e/n), color (e/c), weights (e/n)
#&gt; + edges from 499e405 (vertex names):
#&gt;  [1] RNA_hg19_IL7R  --ADT_CD28             
#&gt;  [2] RNA_hg19_LTB   --ADT_CD28             
#&gt;  [3] RNA_hg19_NKG7  --ADT_CD28             
#&gt;  [4] RNA_hg19_CST7  --ADT_CD28             
#&gt;  [5] RNA_hg19_GNLY  --ADT_CD28             
#&gt;  [6] RNA_hg19_CCL5  --ADT_CD28             
#&gt; + ... omitted several edges</code></pre>
</div>
</div>
<div id="final-notes" class="section level1">
<h1 class="hasAnchor">
<a href="#final-notes" class="anchor"></a>Final notes</h1>
<p>A compiled version of the vignette is available on the <a href="https://github.com/SydneyBioX/BiocAsia2020CiteFuse/articles/workshopTCiteFuse.html">workshop website</a>.</p>
<p>If you have questions that you could not ask during the workshop, feel free to open an issue on the github repository <a href="https://github.com/SydneyBioX/BiocAsia2020CiteFuse/issues">here</a>.</p>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.2 RC (2020-06-20 r78727)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] gridExtra_2.3               mclust_5.4.6               </span>
<span class="co">#&gt;  [3] igraph_1.1.0                DT_0.13                    </span>
<span class="co">#&gt;  [5] scater_1.16.1               ggplot2_3.3.2              </span>
<span class="co">#&gt;  [7] SingleCellExperiment_1.10.1 SummarizedExperiment_1.18.1</span>
<span class="co">#&gt;  [9] DelayedArray_0.14.0         matrixStats_0.56.0         </span>
<span class="co">#&gt; [11] Biobase_2.48.0              GenomicRanges_1.40.0       </span>
<span class="co">#&gt; [13] GenomeInfoDb_1.24.2         IRanges_2.22.2             </span>
<span class="co">#&gt; [15] S4Vectors_0.26.1            BiocGenerics_0.34.0        </span>
<span class="co">#&gt; [17] CiteFuse_1.0.1             </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] Rtsne_0.14                ggbeeswarm_0.6.0         </span>
<span class="co">#&gt;   [3] colorspace_1.4-1          ellipsis_0.3.1           </span>
<span class="co">#&gt;   [5] ggridges_0.5.2            rprojroot_1.3-2          </span>
<span class="co">#&gt;   [7] XVector_0.28.0            BiocNeighbors_1.6.0      </span>
<span class="co">#&gt;   [9] fs_1.4.1                  rstudioapi_0.11          </span>
<span class="co">#&gt;  [11] farver_2.0.3              graphlayouts_0.7.0       </span>
<span class="co">#&gt;  [13] ggrepel_0.8.2             RSpectra_0.16-0          </span>
<span class="co">#&gt;  [15] splines_4.0.2             knitr_1.28               </span>
<span class="co">#&gt;  [17] heatmap.plus_1.3          polyclip_1.10-0          </span>
<span class="co">#&gt;  [19] jsonlite_1.6.1            alluvial_0.1-2           </span>
<span class="co">#&gt;  [21] kernlab_0.9-29            pheatmap_1.0.12          </span>
<span class="co">#&gt;  [23] uwot_0.1.8                ggforce_0.3.1            </span>
<span class="co">#&gt;  [25] ExPosition_2.8.23         compiler_4.0.2           </span>
<span class="co">#&gt;  [27] dqrng_0.2.1               prettyGraphs_2.1.6       </span>
<span class="co">#&gt;  [29] backports_1.1.8           assertthat_0.2.1         </span>
<span class="co">#&gt;  [31] Matrix_1.2-18             limma_3.44.3             </span>
<span class="co">#&gt;  [33] tweenr_1.0.1              BiocSingular_1.4.0       </span>
<span class="co">#&gt;  [35] htmltools_0.5.0           tools_4.0.2              </span>
<span class="co">#&gt;  [37] rsvd_1.0.3                gtable_0.3.0             </span>
<span class="co">#&gt;  [39] glue_1.4.1                GenomeInfoDbData_1.2.3   </span>
<span class="co">#&gt;  [41] reshape2_1.4.4            dplyr_1.0.0              </span>
<span class="co">#&gt;  [43] Rcpp_1.0.4.6              pkgdown_1.5.1.9000       </span>
<span class="co">#&gt;  [45] vctrs_0.3.1               crosstalk_1.1.0.1        </span>
<span class="co">#&gt;  [47] DelayedMatrixStats_1.10.0 ggraph_2.0.3             </span>
<span class="co">#&gt;  [49] xfun_0.15                 stringr_1.4.0            </span>
<span class="co">#&gt;  [51] lifecycle_0.2.0           irlba_2.3.3              </span>
<span class="co">#&gt;  [53] statmod_1.4.34            edgeR_3.30.3             </span>
<span class="co">#&gt;  [55] zlibbioc_1.34.0           MASS_7.3-51.6            </span>
<span class="co">#&gt;  [57] scales_1.1.1              tidygraph_1.2.0          </span>
<span class="co">#&gt;  [59] rhdf5_2.32.1              RColorBrewer_1.1-2       </span>
<span class="co">#&gt;  [61] SNFtool_2.3.0             yaml_2.2.1               </span>
<span class="co">#&gt;  [63] memoise_1.1.0             segmented_1.1-0          </span>
<span class="co">#&gt;  [65] stringi_1.4.6             desc_1.2.0               </span>
<span class="co">#&gt;  [67] randomForest_4.6-14       scran_1.16.0             </span>
<span class="co">#&gt;  [69] BiocParallel_1.22.0       rlang_0.4.6              </span>
<span class="co">#&gt;  [71] pkgconfig_2.0.3           bitops_1.0-6             </span>
<span class="co">#&gt;  [73] evaluate_0.14             lattice_0.20-41          </span>
<span class="co">#&gt;  [75] purrr_0.3.4               Rhdf5lib_1.10.0          </span>
<span class="co">#&gt;  [77] labeling_0.3              htmlwidgets_1.5.1        </span>
<span class="co">#&gt;  [79] cowplot_1.0.0             tidyselect_1.1.0         </span>
<span class="co">#&gt;  [81] plyr_1.8.6                magrittr_1.5             </span>
<span class="co">#&gt;  [83] R6_2.4.1                  generics_0.0.2           </span>
<span class="co">#&gt;  [85] pillar_1.4.4              withr_2.2.0              </span>
<span class="co">#&gt;  [87] mixtools_1.2.0            survival_3.2-3           </span>
<span class="co">#&gt;  [89] RCurl_1.98-1.2            tibble_3.0.1             </span>
<span class="co">#&gt;  [91] crayon_1.3.4              rmarkdown_2.3            </span>
<span class="co">#&gt;  [93] viridis_0.5.1             locfit_1.5-9.4           </span>
<span class="co">#&gt;  [95] grid_4.0.2                FNN_1.1.3                </span>
<span class="co">#&gt;  [97] propr_4.2.6               digest_0.6.25            </span>
<span class="co">#&gt;  [99] tidyr_1.1.0               dbscan_1.1-5             </span>
<span class="co">#&gt; [101] munsell_0.5.0             beeswarm_0.2.3           </span>
<span class="co">#&gt; [103] viridisLite_0.3.0         vipor_0.4.5</span></pre></body></html></div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:yingxin.lin@sydney.edu.au" class="email">yingxin.lin@sydney.edu.au</a><a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p><a href="mailto:hani.kim@sydney.edu.au" class="email">hani.kim@sydney.edu.au</a><a href="#fnref2" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Hani Kim, Yingxin Lin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
