#library("devtools")
#devtools::install("C:/Users/Haibo/ATACseqQCWorkshop", build_vignettes = TRUE)

suppressPackageStartupMessages({
    library(ATACseqQC)
    library(ChIPpeakAnno)
    library(BSgenome.Hsapiens.UCSC.hg38)
    library(TxDb.Hsapiens.UCSC.hg38.knownGene)
    library(MotifDb)
    library(SRAdb)
    library(GenomicAlignments)
})

## prepare the BAM file for importing for Tasks 1 and 2
dupBamFile <- system.file("vignettes/extdata", 
                          "SRR891270.chr20.full.bam", 
                          package = "ATACseqQCWorkshop")
bamFileLabels <- gsub(".full.bam", "", basename(dupBamFile ))

## Task 1: check read alignment statistics using the function`bamQC`

## bamQC
bamQC(dupBamFile, outPath=NULL)

## Task 2: check library complexity
estimateLibComplexity(readsDupFreq(dupBamFile))

## Task 3: plot library insert size distribution
bamFile <- system.file("vignettes/extdata", 
                       "SRR891270.chr20.rmdup.bam", 
                       package = "ATACseqQCWorkshop")
bamFileLabels <- gsub(".rmdup.bam", "", 
                      basename(bamFile))

fragSize <- fragSizeDist(bamFile, bamFileLabels)

## Task 4. plot gloabal signal enrichment around TSSs
## bamfile tags to be read in
possibleTag <- combn(LETTERS, 2)
possibleTag <- c(paste0(possibleTag[1, ], possibleTag[2, ]),
                 paste0(possibleTag[2, ], possibleTag[1, ]))
bamTop100 <- scanBam(BamFile(bamFile, yieldSize = 100),
                     param = ScanBamParam(tag=possibleTag))[[1]]$tag
tags <- names(bamTop100)[lengths(bamTop100)>0]

## files will be output into outPath
outPath <- "splitBam"
if (dir.exists(outPath))
{
    unlink(outPath, recursive = TRUE, force = TRUE)
}
dir.create(outPath)

## shift the coordinates of 5'ends of alignments in the bam file
seqlev <- "chr20" ## subsample data for quick run
which <- as(seqinfo(Hsapiens)[seqlev], "GRanges")
gal <- readBamFile(bamFile, tag=tags, which=which, asMates=TRUE, bigFile=TRUE)
shiftedBamFile <- file.path(outPath, "shifted.bam")
gal1 <- shiftGAlignmentsList(gal, outbam=shiftedBamFile)

## run program for chromosome 1 only
txs <- transcripts(TxDb.Hsapiens.UCSC.hg38.knownGene)
txs <- txs[seqnames(txs) %in% "chr20"]
genome <- Hsapiens

## split the reads into bins for NucleosomeFree, mononucleosome, dinucleosome 
## and trinucleosome. And save the binned alignments into bam files.
objs <- splitGAlignmentsByCut(gal1, txs=txs, genome=genome, outPath = outPath)

## list the files generated by splitGAlignmentsByCut.
dir(outPath)

## get full path to split bam file
bamFiles <- file.path(outPath,
                      c("NucleosomeFree.bam",
                        "mononucleosome.bam",
                        "dinucleosome.bam",
                        "trinucleosome.bam"))
TSS <- promoters(txs, upstream=0, downstream=1)
TSS <- unique(TSS)

## estimate the library size for normalization
librarySize <- estLibSize(bamFiles)

## calculate the signals around TSSs.
NTILE <- 101
dws <- ups <- 1010
sigs <- enrichedFragments(gal=objs[c("NucleosomeFree", 
                                     "mononucleosome",
                                     "dinucleosome",
                                     "trinucleosome")], 
                          TSS=TSS,
                          librarySize=librarySize,
                          seqlev=seqlev,
                          TSS.filter=0.5,
                          n.tile = NTILE,
                          upstream = ups,
                          downstream = dws)

## log2 transformed signals
sigs.log2 <- lapply(sigs, function(.ele) log2(.ele+1))

#plot heatmap
featureAlignedHeatmap(sigs.log2, reCenterPeaks(TSS, width=ups+dws),
                      zeroAt=.5, n.tile=NTILE)

## get signals normalized for nucleosome-free and nucleosome-bound regions.
out <- featureAlignedDistribution(sigs, 
                                  reCenterPeaks(TSS, width=ups+dws),
                                  zeroAt=.5, n.tile=NTILE, type="l", 
                                  ylab="Averaged coverage")

range01 <- function(x){(x-min(x))/(max(x)-min(x))}
out <- apply(out, 2, range01)
matplot(out, type="l", xaxt="n", 
        xlab="Position (bp)", 
        ylab="Fraction of signal")
axis(1, at=seq(0, 100, by=10)+1, 
     labels=c("-1K", seq(-800, 800, by=200), "1K"), las=2)
abline(v=seq(0, 100, by=10)+1, lty=2, col="gray")


## Task 5: streamline IGV snapshot
## Download data from Download data to a location without white spaces in the 
## full path from the folder **subsetted_filtered_BAM** in my Google Drive, 
## by following the shared link: 
## https://drive.google.com/drive/folders/1zGvqnXAIWelgVk9Gw1Kx5ySlCg0iz7AJ?usp=sharing

## get the function IGVSnapshot()
source(system.file("extdata", "IGVSnapshot.R", package = "ATACseqQC"))

## visualize ATAC-seq signal distributions along a few genes
## replace "full/path/to/BAM/files" to the actual path to the BAM files
IGVSnapshot(maxMem="lm", genomeBuild = "hg38", 
            bamFileFullPathOrURLs = "full/path/to/BAM/files", 
            geneNames =c("SPTLC3", "CRLS1", "NELFCD"))
